### PR TITLE
Make exportNotebook work with new reMarkable versions plus other features

### PR DESCRIPTION
**Pull request by jackjackk**
Tested with rM v1.7.2.3 and v1.8.1.1. Below the list of major changes

- update rm2svg to a 0.0.2 version found in another fork which seems to work with v3 lines format (https://github.com/lschwetlick/maxio/tree/master/tools)
- use find_document and rmtgrep to find matching documents, the same functions used in repull.sh from https://github.com/reHackable/scripts (eventually I think all these should be consolidated, as many helper functions would be the same)
- output (and support as input) a .tar.gz file with all the original downloaded file from the rM. the idea is that you might want to process them differently at a later time, while meanwhile you might have deleted the files on the device itself. if the .tar.gz file is used as input, no ssh connection to the rM is needed
- in case of multiple matches, show also the visibleName and not only the uuid and lastModDate to help with the selection
- add svg optimization via svgo if present
- add pdf conversion via alternative methods (e.g. I use inkscape as default, as I thought it gave me the best results)
- add some arguments options, hopefully self-explanatory. I confess one might be related to my workflow alone, which is the askrename. In fact, I tend to leave a default name for my notes (e.g. Notebook32), so that looking at the created pdf and being offered a chance of renaming I can replace that without a second cmd.
- overwrite stroke width and opacity to default values (I was unhappy with the default pdf rendering by rM, which looked nice on the device but uglier in the downloaded pdf)
- increase verbosity (this might be eventually controlled with an arg option

I didn't test all corner cases, plus there is plenty of optimizations possible (e.g. pdf pages w/ no annotations are still converted to svg for the final pdftk combinations) but maybe it's a starting point to keep this useful tool current.

### CODE REVIEW

File: tools/exportNotebook [hunk 1]
```diff
@@ -7,65 +7,255 @@
...
     fi
 else
     rM2svg_cmd="rM2svg"
 fi
 
-# Check if ssh configuration for "remarkable" exists
-grep -Fxq "host remarkable" ~/.ssh/config
-if [ $? -eq 0 ]; then
-    SSH_IP="root@remarkable"
+[ ! -z $COLOR_ANNOTATIONS ] && rM2svg_cmd="$rM2svg_cmd -c"
```
Comment (edupont) [inline tools/exportNotebook]:
```suggestion
[ -n $COLOR_ANNOTATIONS ] && rM2svg_cmd="$rM2svg_cmd -c"
```

File: tools/exportNotebook [hunk 2]
```diff
@@ -7,65 +7,255 @@
...
 
-test -z "$id" && exit 1
+    DOCUMENTS_UUIDS=(${!DOCUMENTS_NAMES[@]})
+    echo "${#DOCUMENTS_UUIDS[@]}"
+    echo "${DOCUMENTS_UUIDS[@]}"
 
-tmpfolder=$(mktemp -d)
+    # Entry not found
+    if [ "${#DOCUMENTS_UUIDS[@]}" -eq 0 ]; then
+        echo "repull: Unable to find document: $path"
```
Comment (edupont) [inline tools/exportNotebook]:
```suggestion
        echo "exportNotebook: Unable to find document: $path"
```

File: tools/exportNotebook [hunk 3]
```diff
@@ -7,65 +7,255 @@
...
-    SSH_IP="root@remarkable"
+[ ! -z $COLOR_ANNOTATIONS ] && rM2svg_cmd="$rM2svg_cmd -c"
+
+tmpfolder=$(mktemp -d)
+echo "Using temp folder '$tmpfolder'"
+
+OUTPUT_UUID=""
+
+if [ -f "$path" ]; then
+    tar -xvzf "$path" -C "$tmpfolder"
```
Comment (edupont) [inline tools/exportNotebook:155]: Did I miss the check where you ensure that $path is a gzipped tarball?
```suggestion
    tar -xvzf "$path" -C "$tmpfolder" || exit 1
```

File: tools/exportNotebook [hunk 4]
```diff
@@ -93,33 +283,96 @@ then
...
+    # debug
+    #cp -v "$f" ~/debug/"$(basename "$f")" 
+    sed -i "s/stroke-width\:[\.0-9]\+/stroke-width:${STROKE_WIDTH}/g;s/opacity\:[\.0-9]\+/opacity:${OPACITY}/g" "$f"
+    [ ! $? -eq 0 ] && echo "Failed to reset stroke-width and opacity in '$f'" && exit 1
+done
+# - optimize svg with svgo
+if command -v "svgo" > /dev/null; then
+    for f in "${tmpfolder}"/foreground*.svg; do
+        svgo -i "$f"
+        [ ! $? -eq 0 ] && echo "Failed to optimize svg '$f'" && exit 1
```
Comment (edupont) [inline tools/exportNotebook]: Is it bad if svgo fails to optimise the svg file? Could the for loop continue and leave that file (and all other failing ones) unoptimised?
```suggestion
        [ ! $? -eq 0 ] && echo "Failed to optimize svg '$f'" && continue
```

File: tools/exportNotebook [hunk 5]
```diff
@@ -93,33 +283,96 @@ then
...
 
-ssh $control_options -O exit ${SSH_IP}
-rm -Rf "${tmpfolder}" "${control_path_dir}"
+if [ -z $KEEP_TEMPDIR ]; then
+    rm -Rf "${tmpfolder}"
+else
+    echo "Left temporary dir '${tmpfolder}'"
+fi
+
+if [ ! -z $ASK_RENAME ]; then
```
Comment (edupont) [inline tools/exportNotebook]:
```suggestion
if [ -n $ASK_RENAME ]; then
```

File: tools/exportNotebook [hunk 6]
```diff
@@ -93,33 +283,96 @@ then
...
+    xdg-open "${filename}.pdf"
+    [ ! -z "$OUTPUT_UUID" ] && echo "Last modified: ${DOCUMENTS_DATES[$OUTPUT_UUID]}"
```
Comment (edupont) [inline tools/exportNotebook]:
```suggestion
    [ -n "$OUTPUT_UUID" ] && echo "Last modified: ${DOCUMENTS_DATES[$OUTPUT_UUID]}"
```

File: tools/exportNotebook [hunk 7]
```diff
@@ -93,33 +283,96 @@ then
...
+    echo
+    read -rp "New name w/o extension, or press Enter to delete files: " NEWFILENAME
+    echo
+    if [ ! -z "$NEWFILENAME" ]; then
```
Comment (edupont) [inline tools/exportNotebook]:
```suggestion
    if [ -n "$NEWFILENAME" ]; then
```

File: tools/exportNotebook [hunk 8]
```diff
@@ -93,33 +283,96 @@ then
...
+        mv -v "${filename}.pdf" "${NEWFILENAME}.pdf"
```
Comment (edupont) [inline tools/exportNotebook]: Make sure you do know when you overwrite existing files
```suggestion
        mv -vi "${filename}.pdf" "${NEWFILENAME}.pdf"
```

File: tools/exportNotebook [hunk 9]
```diff
@@ -93,33 +283,96 @@ then
...
+        mv -v "${filename}.tar.gz" "${NEWFILENAME}.tar.gz"
```
Comment (edupont) [inline tools/exportNotebook]:
```suggestion
        mv -vi "${filename}.tar.gz" "${NEWFILENAME}.tar.gz"
```

File: tools/exportNotebook [hunk 10]
```diff
@@ -93,33 +283,96 @@ then
...
+    rm -Rf "${tmpfolder}"
+else
+    echo "Left temporary dir '${tmpfolder}'"
+fi
+
+if [ ! -z $ASK_RENAME ]; then
+    xdg-open "${filename}.pdf"
+    [ ! -z "$OUTPUT_UUID" ] && echo "Last modified: ${DOCUMENTS_DATES[$OUTPUT_UUID]}"
+    echo
+    read -rp "New name w/o extension, or press Enter to delete files: " NEWFILENAME
```
Comment (edupont) [inline tools/exportNotebook]: The "Enter to delete" action is surprising. When asked for an entry, empty Enter usually accepts the default, or keeps the current value unchanged.

File: tools/rM2svg [hunk 11]
```diff
@@ -1,42 +1,53 @@
 #!/usr/bin/env python3
+#
+# Script for converting reMarkable tablet ".rm" files to SVG image.
+# this works for the new *.rm format, where each page is a seperate file
```
Comment (edupont) [inline tools/rM2svg]:
```suggestion
# this works for the new *.rm format, where each page is a separate file
```

**Review by edupont — COMMENTED** [review]
Thanks a lot for taking the time to share your improvements and corrections. This is very much appreciated!

File: tools/exportNotebook (hunk 10 above)
Comment (jackjackk) [reply tools/exportNotebook]: Good point! I was carried away by optimizing my workflow to the detriment of wider use. Now enter will continue and "!rm" will delete the files.

File: tools/exportNotebook [hunk 12]
```diff
@@ -361,18 +362,18 @@ else
     echo "Left temporary dir '${tmpfolder}'"
 fi
 
-if [ -n $ASK_RENAME ]; then
+if [ -n "$ASK_RENAME" ]; then
     xdg-open "${filename}.pdf"
     [ -n "$OUTPUT_UUID" ] && echo "Last modified: ${DOCUMENTS_DATES[$OUTPUT_UUID]}"
     echo
-    read -rp "New name w/o extension, or press Enter to delete files: " NEWFILENAME
+    read -rp "New name w/o extension, '!rm' to delete files, Enter to skip: " NEWFILENAME
```
Comment (edupont) [inline tools/exportNotebook]: May I suggest a consistency improvement with usual CLI tools, to show the value used in case no entry is made?

```suggestion
    read -rp "New name w/o extension, '!rm' to delete files [${filename}]: " NEWFILENAME
```

**Review by edupont — APPROVED** [review]
Thanks a lot for the PR!

**Comment by jackjackk** [comment]
@edupont thanks a lot for your review! unfortunately I realized the rm2svg does not support the new .lines format, which is version 5. I was testing it with notebooks created before updating the rM to the latest version, hence with a previous .lines format. Do you know where we can find info to update it? also this c++ code https://github.com/ax3l/lines-are-beautiful/ seemed to not be working after a quick test.

**Comment by jackjackk** [comment]
Alright, there was an added 4-byte field, not sure what for, but for the moment at least it produces the expected output on a couple of test pages. eventually we will need to figure out what is that.

File: tools/exportNotebook (hunk 12 above)
Comment (jackjackk) [reply tools/exportNotebook]: fixed, see https://github.com/reHackable/maxio/pull/29/commits/3cb7b765c2107449b434b8af1053ab51583106d0
